sam.c

```diff
@@ -5831,3 +5884,430 @@ void bam_mplp_destructor(bam_mplp_t iter,
+ * reverse flag or modification strand.
+ */
+int bam_mods_at_next_pos(const bam1_t *b, hts_base_mod_state *state,
+                         hts_base_mod *mods, int n_mods) {
+    if (b->core.flag & BAM_FREVERSE) {
+        if (state->seq_pos < 0)
+            return -1;
+    } else {
+        if (state->seq_pos >= b->core.l_qseq)
+            return -1;
+    }
+
+    int i, j, n = 0;
+    unsigned char base = bam_seqi(bam_get_seq(b), state->seq_pos);
+    state->seq_pos++;
+    if (b->core.flag & BAM_FREVERSE)
+        base = seqi_rc[base];
+
+    for (i = 0; i < state->nmods; i++) {
+        if (state->canonical[i] != base && state->canonical[i] != 15/*N*/)
+            continue;
+
+        if (state->MMcount[i]-- > 0)
+            continue;
+
+        char *MMptr = state->MM[i];
+        if (n < n_mods) {
+            mods[n].modified_base = state->type[i];
+            mods[n].canonical_base = seq_nt16_str[state->canonical[i]];
+            mods[n].strand = state->strand[i];
+            mods[n].qual = state->ML[i] ? *state->ML[i] : -1;
+        }
+        n++;
+        if (state->ML[i])
+            state->ML[i] += (b->core.flag & BAM_FREVERSE)
+                ? -state->MLstride[i]
+                : +state->MLstride[i];
+
+        if (b->core.flag & BAM_FREVERSE) {
+            // process MM list backwards
+            char *cp;
+            for (cp = state->MMend[i]-1; cp != state->MM[i]; cp--)
+                if (*cp == ',')
+                    break;
+            state->MMend[i] = cp;
+            if (cp != state->MM[i])
+                state->MMcount[i] = strtol(cp+1, NULL, 10);
+            else
+                state->MMcount[i] = INT_MAX;
+        } else {
+            if (*state->MM[i] == ',')
+                state->MMcount[i] = strtol(state->MM[i]+1, &state->MM[i], 10);
+            else
+                state->MMcount[i] = INT_MAX;
+        }
+
+        // Multiple mods at the same coords.
+        for (j=i+1; j < state->nmods && state->MM[j] == MMptr; j++) {
+            if (n < n_mods) {
+                mods[n].modified_base = state->type[j];
+                mods[n].canonical_base = seq_nt16_str[state->canonical[j]];
+                mods[n].strand = state->strand[j];
+                mods[n].qual = state->ML[j] ? *state->ML[j] : -1;
+            }
+            n++;
+            state->MMcount[j] = state->MMcount[i];
+            state->MM[j]      = state->MM[i];
+            if (state->ML[j])
+                state->ML[j] += (b->core.flag & BAM_FREVERSE)
+                    ? -state->MLstride[j]
+                    : +state->MLstride[j];
+        }
+        i = j-1;
+    }
+
+    return n;
+}
+
+/*
+ * Looks for the next location with a base modification.
+ */
+int bam_next_basemod(const bam1_t *b, hts_base_mod_state *state,
+                     hts_base_mod *mods, int n_mods, int *pos) {
+    if (state->seq_pos >= b->core.l_qseq)
+        return 0;
```

Comment:
I think this is a failure and should return -1 (as described in the header file).